### PR TITLE
[test]:improve coverage for kallsysms.go in pkg/kallsyms

### PR DESCRIPTION
**Pull request by Dhiren-Mhatre**
# [Add test cases to improve kallsyms package coverage from 40% to 73%]

This PR adds several new test cases to improve code coverage for the kallsyms package. 

## How to use
Reviewers can validate this PR by:
1. Running the full test suite: `go test ./pkg/kallsyms/`
2. Verifying the increased coverage: `go tool cover -func=coverage.out`

## Testing done
```bash
$ go test -coverprofile=coverage.out ./pkg/kallsyms/
ok  	github.com/inspektor-gadget/inspektor-gadget/pkg/kallsyms	0.482s
```
image

Part of [Feat:Improve Unit Test Coverage wherever required] #3835

### CODE REVIEW

**Comment by Dhiren-Mhatre** [comment]
@burak-ok  I noticed that my PR checks failed, but I didn’t see any issues in the changes I made. Could you help me confirm if the failure is due to my changes in the test file or if there’s another issue?

File: pkg/kallsyms/kallsyms_test.go [hunk 1]
```diff
@@ -224,3 +226,258 @@ func BenchmarkKernelSymbolAddressPreLoading(b *testing.B) {
...
+		},
+		{
+			name:        "invalid_address",
+			input:       "INVALID_ADDR A fixed_percpu_data",
+			expectedErr: "strconv.ParseUint",
+		},
+		{
+			name:        "malformed_line",
+			input:       "0000000000000000",
+			expectedErr: "line \"0000000000000000\" has less than 3 fields",
```
Comment (burak-ok) [inline pkg/kallsyms/kallsyms_test.go:249]: The error message is the same as the first test case. Is this a different codepath?

File: pkg/kallsyms/kallsyms_test.go [hunk 2]
```diff
@@ -224,3 +226,258 @@ func BenchmarkKernelSymbolAddressPreLoading(b *testing.B) {
...
+		ip       uint64
+		expected string
+	}{
+		{
+			name:     "exact_match_start",
+			ip:       0x0,
+			expected: "first_symbol",
+		},
+		{
+			name:     "before_first_symbol",
```
Comment (burak-ok) [inline pkg/kallsyms/kallsyms_test.go:286]: This is the exact same `ip` as `exact_match_start` and this test is indeed not before the first symbol

File: pkg/kallsyms/kallsyms_test.go [hunk 3]
```diff
@@ -224,3 +226,258 @@ func BenchmarkKernelSymbolAddressPreLoading(b *testing.B) {
...
+		t.Run(tt.name, func(t *testing.T) {
+			reader := strings.NewReader(tt.input)
+			_, err := NewKAllSymsFromReader(reader)
+			require.Error(t, err)
+			require.Contains(t, err.Error(), tt.expectedErr)
+		})
+	}
+}
+
+func TestLookupByInstructionPointerEdgeCases(t *testing.T) {
```
Comment (burak-ok) [inline pkg/kallsyms/kallsyms_test.go:263]: The are some edge cases that are already tested in `TestCustomKAllSymsInstructionPointer`. Could you instead of writing a new function update that?
Especially with the helpful test names you did in this function

File: pkg/kallsyms/kallsyms_test.go [hunk 4]
```diff
@@ -224,3 +226,258 @@ func BenchmarkKernelSymbolAddressPreLoading(b *testing.B) {
...
+		t.Run(tt.name, func(t *testing.T) {
+			reader := strings.NewReader(tt.input)
+			_, _, err := kernelSymbolAddressFromReader(reader, tt.symbol)
+			require.Error(t, err)
+			require.Contains(t, err.Error(), tt.expectedErr)
+		})
+	}
+}
+
+func TestKernelSymbolAddressFromReaderModuleSuccess(t *testing.T) {
```
Comment (burak-ok) [inline pkg/kallsyms/kallsyms_test.go:364]: This kind of test is already done in `TestCustomKAllSymsParsing`

File: pkg/kallsyms/kallsyms_test.go [hunk 5]
```diff
@@ -224,3 +226,258 @@ func BenchmarkKernelSymbolAddressPreLoading(b *testing.B) {
...
+	require.Equal(t, "test_module", mod)
+}
+
+func TestSymbolExistsWithEmptySymbol(t *testing.T) {
+	exists := SymbolExists("")
+	require.False(t, exists)
+}
+
+// Test for parsing various module formats
+func TestKernelSymbolAddressFromReaderModuleFormats(t *testing.T) {
```
Comment (burak-ok) [inline pkg/kallsyms/kallsyms_test.go:379]: These kind of tests are already done in other functions. But IMO I think it is good to have descriptive tests targeting something specific. So I think this is okay

File: pkg/kallsyms/kallsyms_test.go [hunk 6]
```diff
@@ -224,3 +226,258 @@ func BenchmarkKernelSymbolAddressPreLoading(b *testing.B) {
...
+	}{
+		{
+			name:     "no_module",
+			input:    "ffffffffc1234567 T test_symbol",
+			symbol:   "test_symbol",
+			wantMod:  "",
+			wantAddr: 0xffffffffc1234567,
+		},
+		{
+			name:     "with_module",
```
Comment (burak-ok) [inline pkg/kallsyms/kallsyms_test.go:395]: This test is kind of the same as `module_with_with_underscore` since the modile name has an underscore here too

File: pkg/kallsyms/kallsyms_test.go [hunk 7]
```diff
@@ -224,3 +226,258 @@ func BenchmarkKernelSymbolAddressPreLoading(b *testing.B) {
...
+	require.Error(t, err)
+	require.Contains(t, err.Error(), "mock resolver error")
+
+	err = specUpdateAddresses(
+		[]symbolResolver{failResolver},
+		spec,
+		[]string{"symbol1"},
+	)
+	require.Error(t, err)
+	require.Contains(t, err.Error(), "mock resolver error")
```
Comment (burak-ok) [inline pkg/kallsyms/kallsyms_test.go:482]: Since you already mocked the symbol resolver: Could you add a test where the resolving works?

File: pkg/kallsyms/kallsyms_test.go [hunk 8]
```diff
@@ -224,3 +226,258 @@ func BenchmarkKernelSymbolAddressPreLoading(b *testing.B) {
...
+
+	err := specUpdateAddresses(
+		[]symbolResolver{failResolver},
+		spec,
+		[]string{"symbol1"},
+	)
+	require.Error(t, err)
+	require.Contains(t, err.Error(), "mock resolver error")
+
+	err = specUpdateAddresses(
```
Comment (burak-ok) [inline pkg/kallsyms/kallsyms_test.go:476]: Please add a description that this is specifically designed to test if the error cache works. Also would be good to somehow differentiate between the cached error and the non cached one

File: pkg/kallsyms/kallsyms_test.go [hunk 9]
```diff
@@ -198,29 +237,147 @@ func TestRealKAllSymsParsing(t *testing.T) {
 	}
 }
 
-// BenchmarkKernelSymbolAddressJITParsing searches for the symbol while
-// reading the kallsyms file
-func BenchmarkKernelSymbolAddressJITParsing(b *testing.B) {
```
Comment (burak-ok) [inline pkg/kallsyms/kallsyms_test.go:203]: This function is used for benchmarking (see `b *testing.B`). This function shouldn't be removed

File: pkg/kallsyms/kallsyms_test.go [hunk 10]
```diff
@@ -198,29 +237,147 @@ func TestRealKAllSymsParsing(t *testing.T) {
...
+			_, err := NewKAllSymsFromReader(reader)
+			require.Error(t, err)
+			require.Contains(t, err.Error(), tt.expectedErr)
+		})
 	}
 }
 
-// BenchmarkKernelSymbolAddressPreLoading loads the full kallsyms before
-// looking up the symbol
-func BenchmarkKernelSymbolAddressPreLoading(b *testing.B) {
```
Comment (burak-ok) [inline pkg/kallsyms/kallsyms_test.go:215]: This function is used for benchmarking (see `b *testing.B`). This function shouldn't be removed

File: pkg/kallsyms/kallsyms_test.go (hunk 7 above)
Comment (burak-ok) [reply pkg/kallsyms/kallsyms_test.go]: Seems like we would somehow need to create instances `ebpf.VariableSpec`. Let's leave this for later

File: pkg/kallsyms/kallsyms_test.go [hunk 11]
```diff
@@ -198,29 +237,147 @@ func TestRealKAllSymsParsing(t *testing.T) {
...
+
+	t.Run("all_resolvers_fail", func(t *testing.T) {
+		resolver1 := &mockSymbolResolver{
+			resolveFunc: func(symbol string) (uint64, error) {
+				return 0, fmt.Errorf("first resolver failed")
+			},
+		}
+		resolver2 := &mockSymbolResolver{
+			resolveFunc: func(symbol string) (uint64, error) {
+				return 0, fmt.Errorf("second resolver failed")
```
Comment (burak-ok) [inline pkg/kallsyms/kallsyms_test.go:320]: This error message could be defined inside the `t.Run("all_resolvers_fail", func(t *testing.T) {` function as a const. Then the variable can be used in the `.Contains` check below

File: pkg/kallsyms/kallsyms_test.go [hunk 12]
```diff
@@ -198,29 +237,147 @@ func TestRealKAllSymsParsing(t *testing.T) {
...
+			},
+		}
+
+		err := specUpdateAddresses(
+			[]symbolResolver{resolver1, resolver2},
+			spec,
+			[]string{"missing_symbol"},
+		)
+		require.Error(t, err)
+		require.Contains(t, err.Error(), "second resolver failed")
```
Comment (burak-ok) [inline pkg/kallsyms/kallsyms_test.go:336]: Descriptive test case messages would be nice here. Something like `Only the error of the last resolve should be returned`

File: pkg/kallsyms/kallsyms_test.go [hunk 13]
```diff
@@ -198,29 +237,147 @@ func TestRealKAllSymsParsing(t *testing.T) {
...
+
+		// Verify error is cached
+		cachedErr, exists := triedGetAddr["missing_symbol"]
+		require.True(t, exists)
```
Comment (burak-ok) [inline pkg/kallsyms/kallsyms_test.go:340]: Another descriptive test case message here

**Review by burak-ok — COMMENTED** [review]
The tests itself look good. Could you review if you need `.require` instead of `.assert` everywhere? Also the test case messages can be more descriptive

**Comment by Dhiren-Mhatre** [comment]
I've converted require to assert where appropriate and added descriptive messages throughout the test file. Thanks for the review!

**Review by burak-ok — APPROVED** [review]
LGTM. Thanks for the contribution :)
